### PR TITLE
Type scheme once only

### PR DESCRIPTION
**Pull request by ghost**
I would typically have a component assembler in my codebase somewhere that would interact with Crossroads.

The assembler would build paths. 

Consuming the scheme once makes it a bit easier to manage an assembler and I want to avoid typing strings often.

See this pull request in action [here](https://github.com/rob-nash/Poke)

### CODE REVIEW

**Comment by giginet** [comment]
Cool! I'm waiting WWDC keynote now. So I'll check this PR tomorrow 😅

**Comment by ghost** [comment]
Me too! Enjoy.

File: Sources/Crossroad/Router.swift [hunk 1]
```diff
@@ -3,21 +3,13 @@ import Foundation
 public typealias SimpleRouter = Router<Void>
 
 public final class Router<UserInfo> {
-    public let scheme: String
+    private let scheme: String
```
Comment (giginet) [inline Sources/Crossroad/Router.swift:6]: Should we conceal this property?

File: Tests/CrossroadTests/RouterTests.swift [hunk 2]
```diff
@@ -22,6 +22,23 @@ final class RouterTest: XCTestCase {
         XCTAssertFalse(router.responds(to: URL(string: "notfoobar://aaa/bbb")!))
         XCTAssertTrue(router.responds(to: URL(string: "foobar://spam/ham")!))
     }
+    
+    func testCanRespond2() {
```
Comment (giginet) [inline Tests/CrossroadTests/RouterTests.swift]: `testCanRespondWithoutScheme` and others

File: Sources/Crossroad/Router.swift [hunk 3]
```diff
@@ -3,21 +3,13 @@ import Foundation
...
     private var routes: [Route<UserInfo>] = []
 
     public init(scheme: String) {
         self.scheme = scheme
     }
-
-    internal func register(_ route: Route<UserInfo>) {
-        if scheme != route.patternURL.scheme {
-            assertionFailure("Router and pattern must have the same schemes. expect: \(scheme), actual: \(route.patternURL.scheme)")
```
Comment (giginet) [inline Sources/Crossroad/Router.swift]: Please do not remove this error.

File: Tests/CrossroadTests/RouterTests.swift [hunk 4]
```diff
@@ -22,6 +22,23 @@ final class RouterTest: XCTestCase {
...
+        let router = SimpleRouter(scheme: schema)
```
Comment (giginet) [inline Tests/CrossroadTests/RouterTests.swift]: Oh, I found typo. Could you fix them?

`/schema/scheme/`

File: Sources/Crossroad/Router.swift [hunk 5]
```diff
@@ -35,12 +27,17 @@ public final class Router<UserInfo> {
 
     public func register(_ routes: [(String, Route<UserInfo>.Handler)]) {
         for (pattern, handler) in routes {
+            var pattern = pattern
```
Comment (giginet) [inline Sources/Crossroad/Router.swift]: You can write more clearly code.

```swift
    public func register(_ routes: [(String, Route<UserInfo>.Handler)]) {
        for (pattern, handler) in routes {
            let patternURLString: String
            if pattern.hasPrefix("\(scheme)://") {
                patternURLString = pattern
            } else {
                patternURLString = "\(scheme)://\(pattern)"
            }
            guard let patternURL = PatternURL(string: patternURLString) else {
                assertionFailure("\(pattern) is invalid")
                continue
            }
            let route = Route(pattern: patternURL, handler: handler)
            register(route)
        }
    }
```

Comment (giginet) [reply Sources/Crossroad/Router.swift]: I think you needn't remove `internal func register(_ route: Route<UserInfo>)`.

File: Tests/CrossroadTests/RouterTests.swift [hunk 6]
```diff
@@ -22,6 +22,23 @@ final class RouterTest: XCTestCase {
...
+            ("foo/bar", { _ in true }),
+            ("spam/ham", { _ in false }),
+            (":keyword", { _ in true }),
+            ("foo/:keyword", { _ in true }),
+            ])
+        XCTAssertTrue(router.responds(to: URL(string: "foobar://static")!))
+        XCTAssertTrue(router.responds(to: URL(string: "foobar://foo")!))
+        XCTAssertTrue(router.responds(to: URL(string: "foobar://foo/bar")!))
+        XCTAssertTrue(router.responds(to: URL(string: "foobar://foo/10000")!))
+        XCTAssertFalse(router.responds(to: URL(string: "notfoobar://aaa/bbb")!))
```
Comment (giginet) [inline Tests/CrossroadTests/RouterTests.swift:45]: Could you add assertions to check without scheme patterns?

```swift
XCTAssertFalse(router.responds(to: URL(string: "static")!))
XCTAssertFalse(router.responds(to: URL(string: "foo/bar")!))
XCTAssertFalse(router.responds(to: URL(string: "spam/ham")!))
```

**Review by giginet — COMMENTED** [review]
@rob-nash Hi!
Thank you contribution. Your proposal looks good. however, some points should be fixed.

File: Sources/Crossroad/Router.swift (hunk 1 above)
Comment (giginet) [reply Sources/Crossroad/Router.swift:6]: Maybe, I guess that nobody use `Router.scheme` from external. 🤔

Comment (ghost) [reply Sources/Crossroad/Router.swift:6]: I think so. I think it might be good to release API, once it becomes useful.

**Comment by ghost** [comment]
Have made the changes @giginet suggested 👍

File: Sources/Crossroad/Router.swift [hunk 7]
```diff
@@ -3,21 +3,21 @@ import Foundation
...
 public final class Router<UserInfo> {
-    public let scheme: String
+    private let scheme: String
     private var routes: [Route<UserInfo>] = []
 
     public init(scheme: String) {
         self.scheme = scheme
     }
-
+
```
Comment (giginet) [inline Sources/Crossroad/Router.swift]: Could you execute `swiftlint autocorrect`?

**Review by giginet — APPROVED** [review]
Great job! 👍 

I'll merge this after correcting lint warnings.

https://travis-ci.org/giginet/Crossroad/jobs/388382200#L604

Thank you! 🎉

**Comment by giginet** [comment]
You can ignore `type_body_length` violation. 
But, please fix `trailing_whitespace` violations.
You will fix them by only executing `swiftlint autocorrect`

**Pull request merged**